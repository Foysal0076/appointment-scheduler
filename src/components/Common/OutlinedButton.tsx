import { cva } from 'class-variance-authority'
import clsx from 'clsx'

import Spinner from '@/components/Common/Spinner'

type ButtonProps = React.ButtonHTMLAttributes<HTMLButtonElement> & {
  color?: 'primary' | 'secondary' | 'tertiary' | 'danger' | 'success'
  size?: 'xs' | 'sm' | 'md' | 'lg' | 'xl'
  fullWidth?: boolean
  loading?: boolean
}

const ButtonVariants = cva(
  'rounded bg-none border-2 focus:ring-4 disabled:active:transform-none disabled:cursor-not-allowed dark:hover:bg-neutral-20/50 transition-all duration-300 text-nowrap focus:outline-none font-medium active:scale-[.9]',
  {
    variants: {
      color: {
        primary:
          'hover:bg-primary-50 focus:ring-primary-100 border-primary-500/80 text-primary-500  focus:ring-primary-100 dark:text-primary-200 dark:border-primary-200 disabled:text-primary-200 disabled:border-primary-100 disabled:hover:bg-transparent dark:disabled:bg-primary-50/20 dark:disabled:border-primary-50/10 dark:disabled:text-primary-100/60',
        secondary:
          'hover:bg-secondary-50 focus:ring-secondary-100 border-secondary-500/80 text-secondary-500  focus:ring-secondary-100 dark:text-secondary-200 dark:border-secondary-200 disabled:text-secondary-200 disabled:border-secondary-100 disabled:hover:bg-transparent dark:disabled:bg-secondary-50/20 dark:disabled:border-secondary-50/10 dark:disabled:text-secondary-100/60',
        tertiary:
          'hover:bg-tertiary-50 focus:ring-tertiary-100 border-tertiary-500/80 text-tertiary-500  focus:ring-tertiary-100 dark:text-tertiary-200 dark:border-tertiary-200 disabled:text-tertiary-200 disabled:border-tertiary-100 disabled:hover:bg-transparent dark:disabled:bg-tertiary-50/20 dark:disabled:border-tertiary-50/10 dark:disabled:text-tertiary-100/60',
        danger:
          'hover:bg-danger-50 focus:ring-danger-100 border-danger-500/80 text-danger-500  focus:ring-danger-100 dark:text-danger-200 dark:border-danger-200 disabled:text-danger-200 disabled:border-danger-100 disabled:hover:bg-transparent dark:disabled:bg-danger-50/20 dark:disabled:border-danger-50/10 dark:disabled:text-danger-100/60',
        success:
          'hover:bg-success-50 focus:ring-success-100 border-success-500/80 text-success-500  focus:ring-success-100 dark:text-success-200 dark:border-success-200 disabled:text-success-200 disabled:border-success-100 disabled:hover:bg-transparent dark:disabled:bg-success-50/20 dark:disabled:border-success-50/10 dark:disabled:text-success-100/60',
      },
      size: {
        xs: 'h-[1.75rem] text-xs px-4',
        sm: 'h-[2rem] text-sm px-4',
        md: 'h-[2.5rem] text-md px-4',
        lg: 'h-[3.125rem] text-lg px-4',
        xl: 'h-[3.5rem] text-lg px-4',
      },
    },
    defaultVariants: {
      color: 'primary',
      size: 'md',
    },
  }
)

const OutlinedButton = ({
  color = 'primary',
  size = 'md',
  loading = false,
  className = '',
  children,
  ...buttonProps
}: ButtonProps) => {
  return (
    <button
      className={clsx(
        `relative select-none`,
        `${ButtonVariants({ color, size })}`,
        `${className}`
      )}
      disabled={loading}
      {...buttonProps}>
      {loading && (
        <div className='absolute inset-0 flex items-center justify-center'>
          <Spinner className='text-current' />
        </div>
      )}
      <div
        className={clsx(`visible font-Rubik`, {
          invisible: loading,
          'opacity-0': loading,
        })}>
        {children}
      </div>
    </button>
  )
}

export default OutlinedButton
