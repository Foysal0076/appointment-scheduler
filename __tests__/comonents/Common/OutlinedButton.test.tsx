import OutlinedButton from '@/components/Common/OutlinedButton'
import { render } from '@testing-library/react'

describe('OutlinedButton component', () => {
  test('renders without crashing', () => {
    const { container } = render(<OutlinedButton />)
    expect(container).toBeInTheDocument()
  })

  test('applies default color and size classes if not provided', () => {
    const { container } = render(<OutlinedButton />)
    const buttonElement = container.querySelector('button')
    expect(buttonElement).toHaveClass('border-primary-500/80')
    expect(buttonElement).toHaveClass('h-[2.5rem]')
  })

  test('applies provided color and size classes', () => {
    const { container } = render(<OutlinedButton color='secondary' size='lg' />)
    const buttonElement = container.querySelector('button')
    expect(buttonElement).toHaveClass('border-secondary-500/80')
    expect(buttonElement).toHaveClass('h-[3.125rem]')
  })

  test('disables the button when loading is true', () => {
    const { container } = render(<OutlinedButton loading />)
    const buttonElement = container.querySelector('button')
    expect(buttonElement).toBeDisabled()
  })

  test('renders spinner when loading is true', () => {
    const { container } = render(<OutlinedButton loading />)
    const spinnerElement = container.querySelector('.absolute')
    expect(spinnerElement).toBeInTheDocument()
  })

  test('renders children when not loading', () => {
    const { container } = render(<OutlinedButton>Click me</OutlinedButton>)
    const childrenElement = container.querySelector('.visible')
    expect(childrenElement).toBeInTheDocument()
    expect(childrenElement).toHaveTextContent('Click me')
  })

  test('matches snapshot', () => {
    const { asFragment } = render(<OutlinedButton />)
    expect(asFragment()).toMatchSnapshot()
  })
})
