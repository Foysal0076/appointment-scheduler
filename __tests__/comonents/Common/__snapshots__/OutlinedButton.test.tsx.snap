// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OutlinedButton component matches snapshot 1`] = `
<DocumentFragment>
  <button
    class="relative select-none rounded bg-none border-2 focus:ring-4 disabled:active:transform-none disabled:cursor-not-allowed dark:hover:bg-neutral-20/50 transition-all duration-300 text-nowrap focus:outline-none font-medium active:scale-[.9] hover:bg-primary-50 focus:ring-primary-100 border-primary-500/80 text-primary-500  focus:ring-primary-100 dark:text-primary-200 dark:border-primary-200 disabled:text-primary-200 disabled:border-primary-100 disabled:hover:bg-transparent dark:disabled:bg-primary-50/20 dark:disabled:border-primary-50/10 dark:disabled:text-primary-100/60 h-[2.5rem] text-md px-4"
  >
    <div
      class="visible font-Rubik"
    />
  </button>
</DocumentFragment>
`;
